From cbf4723f9bfd3e998b07d8129f4188ba49b76dd0 Mon Sep 17 00:00:00 2001
From: Mathias Biilmann Christensen <mbc@mbcbook-2.local>
Date: Thu, 18 Mar 2010 01:15:52 +0100
Subject: [PATCH] Deleting and setting a cookie in the same request was broken

Made sure to remove a cookie from @deleted_cookies when set
---
 .../lib/action_dispatch/middleware/cookies.rb      |    1 +
 actionpack/test/controller/cookie_test.rb          |   11 +++++++++++
 2 files changed, 12 insertions(+), 0 deletions(-)

diff --git a/actionpack/lib/action_dispatch/middleware/cookies.rb b/actionpack/lib/action_dispatch/middleware/cookies.rb
index 0dc03a1..ab7130a 100644
--- a/actionpack/lib/action_dispatch/middleware/cookies.rb
+++ b/actionpack/lib/action_dispatch/middleware/cookies.rb
@@ -84,6 +84,7 @@ module ActionDispatch
 
         options[:path] ||= "/"
         @set_cookies[key] = options
+        @delete_cookies.delete(key)
         value
       end
 
diff --git a/actionpack/test/controller/cookie_test.rb b/actionpack/test/controller/cookie_test.rb
index 908967a..36498d1 100644
--- a/actionpack/test/controller/cookie_test.rb
+++ b/actionpack/test/controller/cookie_test.rb
@@ -64,6 +64,12 @@ class CookieTest < ActionController::TestCase
       cookies.permanent.signed[:remember_me] = 100
       head :ok
     end
+
+    def delete_and_set_cookie
+      cookies.delete :user_name
+      cookies[:user_name] = { :value => "david", :expires => Time.utc(2005, 10, 10,5) }
+      head :ok
+    end
   end
 
   tests TestController
@@ -152,6 +158,11 @@ class CookieTest < ActionController::TestCase
     assert_equal 100, @controller.send(:cookies).signed[:remember_me]
   end
 
+  def test_delete_and_set_cookie
+    get :delete_and_set_cookie
+    assert_cookie_header "user_name=david; path=/; expires=Mon, 10-Oct-2005 05:00:00 GMT"
+    assert_equal({"user_name" => "david"}, @response.cookies)
+  end
 
   private
     def assert_cookie_header(expected)
-- 
1.6.3.1

